Add tensor centralities

nwlandry opened the pull request:
Added tensor methods to the centrality module. Fixes #505.

This PR also fixes Issue #616, where the `node_edge_centrality` was not converging correctly.

nwlandry commented:
@ilyaamburg --- I rewrote the `pairwise_incidence` method and I have a few questions on this: (1) does order matter when adding (i, j) to the dictionary of pairs? (2) there was an "r" parameter for this function, but because `r` seemed to just be the maximum edge size, the loop where you added the self-loops (node, node) seems to always be true. So I removed `r` as an argument, and removed the if-statement. Let me know your thoughts!

maximelucas commented on xgi/utils/tensor.py:
```diff
@@ -0,0 +1,191 @@
+    return gen_fun_coef
+
+
+def ttsv1(E, H, r, a):
```
missing docs. also the name is not very explicit, anyway of finding another one that's not too long?

ilyaamburg replied:
I will add docs. It is "tensor times same vector in all modes but one" -- ttsv is standard in the tensor literature/software world, so maybe we're ok?

nwlandry replied:
Done.

maximelucas commented on xgi/utils/tensor.py:
```diff
@@ -0,0 +1,191 @@
+    return s
+
+
+def ttsv2(E, H, r, a, n):
```
same as ttsv1

ilyaamburg replied:
Yes.

maximelucas commented on xgi/utils/tensor.py:
```diff
@@ -0,0 +1,191 @@
+    return sum(((-1) ** j) * binomial(l, j) * (l - j) ** r for j in range(l + 1))
+
+
+def get_gen_coef_subset_expansion(edge_values, node_value, r):
```
missing docs also a few functions after this one

ilyaamburg replied:
I will add these.

nwlandry replied:
I added an underscore to emphasize that we're not exposing these functions to users. I moved them to the end of the file and added a note that they are helper functions. TBH, I'm not the best person to write docstrings for these functions, but if Ilya wants to give it a go, I'm fine with that.

maximelucas commented on xgi/utils/tensor.py:
```diff
@@ -0,0 +1,191 @@
+    return pairs
+
+
+def banerjee_coeff(l, r):
+    """Return the Banerjee alpha coefficient
```
is this a standard coefficient? googling didn't help me. Either change the name, or say what it relates to, and maybe add a ref?

ilyaamburg replied:
We should add reference to the Banerjee paper https://www.sciencedirect.com/science/article/pii/S0024379516306115?via%3Dihub

nwlandry replied:
Reference added.

maximelucas commented on xgi/utils/tensor.py:
```diff
@@ -0,0 +1,191 @@
+
+    Parameters
+    ----------
+    l : int
+        size of given hyperedge
+    r : int
```
should we replace these by "s" or "size" and "s_max" to make it more explicit and consistent with the rest of the library?

ilyaamburg replied:
I'm down with "size"!

nwlandry replied:
Fixed.

maximelucas commented on xgi/utils/tensor.py:
```diff
@@ -0,0 +1,191 @@
+]
+
+
+def pairwise_incidence(edgedict):
+    """Create pairwise adjacency dictionary from hyperedge list dictionary
```
Function name says "incidence" but description says "adjacency", maybe pick one. 

Also, this feel like it could/should be a general function? If I got it right, we have `xgi.adjacency_matrix(H, order=1)` that associates a pair of nodes to the number of edges of which it's part, and here we associated it to the list of edge IDs, with a dict. Would it make sense to have a new one like `xgi.adjacency_dict(H, order)` somewhere (not in utils/tensor.py)?

ilyaamburg replied:
You're right, we should stick with one convention, and it should be incidence. Yeah, it might make sense to have to have an xgi.incidence_dict(H, order) function.

nwlandry replied:
After looking more into this, I think that it's best to stay in the `tensor` utilities module. Updating the docstring right now.

maximelucas commented on xgi/algorithms/centrality.py:
```diff
@@ -392,3 +387,161 @@ def katz_centrality(H, cutoff=100):
+
+
+def h_eigenvector_tensor_centrality(H, max_iter=100, tol=1e-6):
+    """Compute the H-eigenvector centrality of a hypergraph.
```
maybe add one line of additional info saying for example what H stands for, or what the specificity of this centrality is?

ilyaamburg replied:
I'm thinking of adding a reference for H- (and Z-) eigenvectors of tensors.

nwlandry replied:
Done.

maximelucas commented on xgi/algorithms/centrality.py:
```diff
@@ -392,3 +387,161 @@ def katz_centrality(H, cutoff=100):
+
+    See Also
+    --------
+    clique_eigenvector_centrality
```
```suggestion
    clique_eigenvector_centrality
    z_eigenvector_tensor_centrality
```

maximelucas commented on xgi/algorithms/centrality.py:
```diff
@@ -392,3 +387,161 @@ def katz_centrality(H, cutoff=100):
+
+    See Also
+    --------
+    clique_eigenvector_centrality
```
```suggestion
    clique_eigenvector_centrality
    h_eigenvector_tensor_centrality
```

maximelucas commented on xgi/algorithms/centrality.py:
```diff
@@ -392,3 +387,161 @@ def katz_centrality(H, cutoff=100):
+            evec = -evec
+        return evec / norm(evec, 1)
+
+    def f(u):
```
what does this do?

ilyaamburg replied:
This makes the eigenvector positive if you happen to compute a negative one and makes it stochastic.

maximelucas commented:
Thanks Nich. A few comments mainly about documentation. Let's see what @ilyaamburg has to say about your question too.

ilyaamburg commented:
> @ilyaamburg --- I rewrote the `pairwise_incidence` method and I have a few questions on this: (1) does order matter when adding (i, j) to the dictionary of pairs? (2) there was an "r" parameter for this function, but because `r` seemed to just be the maximum edge size, the loop where you added the self-loops (node, node) seems to always be true. So I removed `r` as an argument, and removed the if-statement. Let me know your thoughts!

Hey Nick, I am liking the new pairwise_incidence that uses combinations and defaultdict with set values, which should be fine, since the order does not matter as long as you end up one of each pair, and the order of the values does not matter either. The r and corresponding if statement should stay, though, because we only want self loops in the case l (= |e|) < r; that is, we do not want to add them if the hyperedge size is maximal.

maximelucas commented on tests/algorithms/test_centrality.py:
```diff
@@ -195,3 +167,125 @@ def test_katz_centrality(edgelist1, edgelist8):
+        assert np.allclose(c[i], true_c[i])
+
+
+def _ratio(r, m, kind="CEC"):
```
should this be in tests or in some utility file?

nwlandry replied:
Maybe? It really is only used to test centralities for the sunflower hypergraph, but since we only use it here, it didn't seem to make sense to move it.

nwlandry commented:
@ilyaamburg --- I think I addressed all of your comments if you want to take a final look. From looking at the test coverage, it seems like the `_get_gen_coef_fft_fast_array` function is never called with the test hypergraphs that I used. Is there a specific hypergraph that I could use to test this?

Also, could you provide docstrings in this comment thread briefly describing the two helper functions, `_get_gen_coef_subset_expansion` and `_get_gen_coef_fft_fast_array`?

We are almost there!!

ilyaamburg commented:
> @ilyaamburg --- I think I addressed all of your comments if you want to take a final look. From looking at the test coverage, it seems like the `_get_gen_coef_fft_fast_array` function is never called with the test hypergraphs that I used. Is there a specific hypergraph that I could use to test this?
> 
> Also, could you provide docstrings in this comment thread briefly describing the two helper functions, `_get_gen_coef_subset_expansion` and `_get_gen_coef_fft_fast_array`?
> 
> We are almost there!!

@nwlandry thanks so much for this! Hm, _get_gen_coef_fft_fast_array is really the workhorse that makes everything scale so I'm shocked it's not being called. It should get called whenever the hyperedge size is in the purple range:
<img width="630" alt="Screenshot 2024-11-12 at 10 15 04 AM" src="https://github.com/user-attachments/assets/da692d69-60e3-4fd0-a36a-a5e49af60884">
As for the docstrings, here is a draft:
def _get_gen_coef_subset_expansion(edge_values, node_value, r):    
    """Computes the generating funciton coefficient of order r using subset expansion.

    Parameters
    ----------
    edge_values : NumPy array
        Array of values from the a vector corresponding to
        nodes in the hyperedge.
    node_value : float
        The value in a corresponding to the node being processed.
    r : int
        Desired order to get coefficient for.

    Returns
    -------
    float
        Generating function coefficient of order r.

    See Also
    --------
    _get_gen_coef_fft_fast_array

    References
    ----------
    Scalable Tensor Methods for Nonuniform Hypergraphs,
    Sinan Aksoy, Ilya Amburg, Stephen Young,
    https://doi.org/10.1137/23M1584472
    """
def _get_gen_coef_fft_fast_array(edge_without_node, a, node, l, r):    
    """Computes the generating funciton coefficient of order r using the Fast Fourier Transform.

    Parameters
    ----------
    edge_without_node : list
        Array of node indices corresponding to
        all nodes in the hyperedge but the one being processed.
    a : NumPy array
        The vector to multiply the tensor by.
    node : int
        The index of the node being processed.
    l : int
        Number of nodes in the hyperedge.
    r : int
        Desired order to get coefficient for.

    Returns
    -------
    float
        Generating function coefficient of order r.

    See Also
    --------
    _get_gen_coef_subset_expansion

    References
    ----------
    Scalable Tensor Methods for Nonuniform Hypergraphs,
    Sinan Aksoy, Ilya Amburg, Stephen Young,
    https://doi.org/10.1137/23M1584472
    """

nwlandry commented:
Everything is now addressed, and so I will now merge this once tests pass.
